Content.Server/Singularity/EntitySystems/ContainmentFieldGeneratorSystem.cs

```diff
@@ -1,264 +1,387 @@
+                ChangeOnLightVisualizer(value.Item1);
             }
+
+            ChangeFieldVisualizer(value.Item1);
         }
+        component.Connections.Clear();
+        component.IsConnected = false;
+        ChangeOnLightVisualizer(component);
+        ChangeFieldVisualizer(component);
+        _popupSystem.PopupEntity(Loc.GetString("comp-containment-disconnected"), component.Owner, Filter.Pvs(component.Owner), PopupType.LargeCaution);
+    }
 
-        private void HandleGeneratorCollide(EntityUid uid, ContainmentFieldGeneratorComponent component, StartCollideEvent args)
+    #endregion
+
+    #region Connections
+
+        /// <summary>
+    /// Stores power in the generator. If it hits the threshold, it tries to establish a connection.
+    /// </summary>
+    /// <param name="power">The power that this generator received from the collision in <see cref="HandleGeneratorCollide"/></param>
+    public void ReceivePower(int power, ContainmentFieldGeneratorComponent component)
+    {
+        component.PowerBuffer += power;
+
+        if (component.PowerBuffer >= component.PowerMinimum)
         {
-            if (_tags.HasTag(args.OtherFixture.Body.Owner, "EmitterBolt"))
+            for (int i = 0; i < 8; i+=2)
             {
-                ReceivePower(6, component);
+                var dir = (Direction)i;
+
+                if (component.Connections.ContainsKey(dir))
+                    continue; // This direction already has an active connection
+
+                TryGenerateFieldConnection(dir, component);
             }
         }
 
-        private void HandleFieldCollide(EntityUid uid, ContainmentFieldComponent component, StartCollideEvent args)
+        ChangePowerVisualizer(power, component);
+    }
+
+    public void LosePower(int power, ContainmentFieldGeneratorComponent component)
+    {
+        component.PowerBuffer -= power;
+
+        if (component.PowerBuffer < component.PowerMinimum && component.Connections.Count != 0)
         {
-            if (component.Parent == null)
-            {
-                EntityManager.QueueDeleteEntity(uid);
-                return;
-            }
+            RemoveConnections(component);
+        }
+
+        ChangePowerVisualizer(power, component);
+    }
+
+    /// <summary>
+    /// This will attempt to establish a connection of fields between two generators.
+    /// If all the checks pass and fields spawn, it will store this connection on each respective generator.
+    /// </summary>
+    /// <param name="dir">The field generator establishes a connection in this direction.</param>
+    /// <param name="component">The field generator component</param>
+    /// <returns></returns>
+    private bool TryGenerateFieldConnection(Direction dir, ContainmentFieldGeneratorComponent component)
+    {
+        if (!component.Enabled) return false;
+
+        var genXForm = Transform(component.Owner);
+        if (!genXForm.Anchored) return false;
```

Comment:
Resolving components is ""expensive"" so really you should pass in transformcomponent as this method is gonna get called 8 times.